### PR TITLE
Update typescript document to match with current types

### PR DESCRIPTION
**Pull request by Ailrun**
**What**: Update typescript document

**Why**: Typings are changed, so document is outdated.

**Checklist**:
- [x] Documentation
- [N/A] Tests
- [N/A] Code complete

This document depends on #678 and #680, so please do not merge this before them.

### CODE REVIEW

File: docs/typescript.md [hunk 1]
```diff
@@ -2,9 +2,77 @@
...
-### html elements
+const bodyStyle = css({
+  display: 'flex',
+  flowDirection: 'column-reverse',
+});
+```
+
+## create-emotion-styled
+
+Current typing for `create-emotion-styled` is only compatible with React, and will not work with Preact. For detail typing, see following `react-element` section.
```
Comment (mitchellhamilton) [inline docs/typescript.md]: Could you change "Current typing for `create-emotion-styled` is" to "The current typings for `create-emotion-styled` are only"?

File: docs/typescript.md [hunk 2]
```diff
@@ -2,9 +2,77 @@
 title: "Typescript"
 ---
 
-Emotion includes TypeScript definitions for `styled` components and has type inferences for both html elements and React components.
+Emotion includes TypeScript definitions for `create-emotion`, `emotion`, `create-emotion-styled`, and `react-emotion` packages. These definition also could infer types for css property (with object syntax), and HTML/SVG tag name, and props types.
```
Comment (mitchellhamilton) [inline docs/typescript.md]: Is there a circumstance where it can't infer types for css properties?
If not, could you replace "These definition also could infer types for css property" with "These definitions also infer types for css properties"?

Also, could you replace "HTML/SVG tag name" with "HTML/SVG tag names"?

Comment (Ailrun) [reply docs/typescript.md]: Allright. I will change them.

File: docs/typescript.md [hunk 3]
```diff
@@ -2,9 +2,77 @@
...
+});
+
+const subtitleStyle = myEmotion.css`
+  boxSizing: 'border-box';
+  width: 100;
+  height: 60;
+`;
+```
+
+Typescript checks css properties in object style syntax using csstype package, so following code will emit errors.
```
[the remark this answers is not in the document]
Comment (Ailrun) [reply docs/typescript.md]: But it would be helpful when users meet some type issues.... wouldn't it?

File: docs/typescript.md (hunk 1 above)
Comment (Ailrun) [reply docs/typescript.md]: OK. Thank you for your review!

File: docs/typescript.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ailrun) [reply docs/typescript.md]: Oh... you're right. I will add the link too.

File: docs/typescript.md [hunk 4]
```diff
@@ -184,3 +225,30 @@ const Button = styled('button')`
 
 export default Button
 ```
+
+## preact-emotion
+
+Types are almost same with `react-emotion` package.
```
Comment (mitchellhamilton) [inline docs/typescript.md]: If they're almost the same and not the exact same, could you explain how they're different?

File: docs/typescript.md [hunk 5]
```diff
@@ -184,3 +225,30 @@ const Button = styled('button')`
...
+
+const bodyStyle = myEmotion.css({
+  display: 'flex',
+  flowDirection: 'column-reverse',
+});
+```
+
+## create-emotion-styled
+
+The current typings for `create-emotion-styled` are only compatible with React, and will not work with Preact. For detail typing, see `react-element` section above.
```
Comment (mitchellhamilton) [inline docs/typescript.md]: This isn't true anymore about not supporting Preact, right?

File: docs/typescript.md [hunk 6]
```diff
@@ -131,33 +169,36 @@ const Component: SFC = ({ label, className }) => (
...
 
-Unfortunately, when you pass custom props to a styled component, TypeScript will stop inferring your Component props, and you will need to re-type them.
-
 ### Define a Theme
 
 By default, the `props.theme` has `any` type annotation and works without error.\
 However, you can define a theme type by creating a another `styled` instance.
 
-_styled.tsx_
+_styled.jsx_
```
Comment (mitchellhamilton) [inline docs/typescript.md]: Why the change in extension, isn't `tsx` generally used?

File: docs/typescript.md (hunk 4 above)
Comment (Ailrun) [reply docs/typescript.md]: OK. I will add following.
> However, this typing use Preact component types like `ComponentConstructor` and `FunctionalComponent` instead of React component types.

Is this enough?

File: docs/typescript.md (hunk 6 above)
Comment (Ailrun) [reply docs/typescript.md]: Oh, my mistake. Sorry. I will change these.

File: docs/typescript.md (hunk 5 above)
Comment (Ailrun) [reply docs/typescript.md]: FYI, Preact and React both modify `JSX.Element`, so we cannot import both typings in one file. That's why I did not provide typing for preact version of `create-emotion-styled`. (Its main typing file already import React typing)

Comment (Ailrun) [reply docs/typescript.md]: The only way I can image, to deal with the problem, is, providing index typing without import any (P)React related typings and making users import them.

However, this may confuse users, because it will not work properly when users do not import (P)React related typings.

File: docs/typescript.md [hunk 7]
```diff
@@ -47,48 +84,42 @@ const App = () => <Link href="#">Click me</Link>
...
 
 type ImageProps = {
   src: string,
+  width: number;
 }
 
-const Image = styled<ImageProps, 'div'>('div')`
-  background: url(${props => props.src}) center center;
+const Image0 = styled('div')`
+  width: ${(props: ImageProps) => props.width};
```
Comment (SpaceK33z) [inline docs/typescript.md]: Is there a reason you are not using `styled<ImageProps, 'div'>('div')`? I think that looks way clearer, especially because then you don't have to do `(props: ImageProps) => ...` for every prop function.

File: docs/typescript.md [hunk 8]
```diff
@@ -47,48 +84,42 @@ const App = () => <Link href="#">Click me</Link>
...
+// Or with generic type
 
-// Or shorthand
-
-const Image = styled.div<ImageProps>({
-  backgroundSize: contain;
-}, ({ src }) => ({
-  background: `url(${src}) center center`,
-}))
+const Image1 = styled('div')<ImageProps>({
```
Comment (SpaceK33z) [inline docs/typescript.md]: This doesn't work, since TS expects a second argument, `styled('div')<ImageProps, 'div'>. TypeScript 2.9 has a fix for this, but 2.9 has not yet been released.

File: docs/typescript.md (hunk 7 above)
Comment (SpaceK33z) [inline docs/typescript.md]: Why are you using `width: ${(props: ImageProps)` instead of typing the component with `styled<ImageProps>('div')`? This is a bit confusing for me, since I see the latter as a much clearer way to type your component.

Comment (Ailrun) [reply docs/typescript.md]: Because with updated typing, there're no type parameters Props for `styled`.

Comment (Ailrun) [reply docs/typescript.md]: In TS 2.9 (which is in RC), you can use template string generic, however, before it, you should give a type to function and let TS infer correct type for component.

Comment (SpaceK33z) [reply docs/typescript.md]: Why? Typing every prop function with `${(props: ImageProps)` is more work, right? The redundant `div` in `styled<ImageProps, 'div'>` is a bit frustrating, but it will not be necessary anymore as soon as TypeScript 2.9 is released.

Comment (SpaceK33z) [reply docs/typescript.md]: > what you mentioned occurs only when you use more than two function argument, and as far as I experienced, it's not that common

In my experience I have a lot of these functions, if you are building a component library you can easily have ten of those prop functions. But if this isn't relevant anymore with TS 2.9 I digress.

Comment (Ailrun) [reply docs/typescript.md]: Maybe my opinion is from my **object syntax** based experience. Even when I make a library, I rarely use more than two function arguments.

Comment (Ailrun) [reply docs/typescript.md]: And, thank you for your feedback!

**Comment by felipeleusin** [comment]
Hi folks,

I just hit an issue. I have a button component that uses custom props but fallsback to theme props. But if I use 

(props: ButtonProps) => whatever

I lose typing to Themed. I've been trying to get this to work and failing. Do you have any pointers? Should I make something Themed<ButtonProps> for this cases? Maybe it's a good thing to add in the docs

**Comment by Ailrun** [comment]
@felipeleusin If you use TS >= 2.9, you can use something like
```
styled('div')<ButtonProps>`
  ${props => something}
`;
```

**Comment by felipeleusin** [comment]
Oh, ok! Will try this. Thanks a lot!

**Comment by Ailrun** [comment]
I should add that to the document...

**Comment by felipeleusin** [comment]
I can add a PR for that, was just checking it's working first :)

**Comment by Ailrun** [comment]
@felipeleusin Oh, I'm not start yet. I'm discussing about typing in #721.
